In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [3]:
ath = r'/Users/ECC/Career Foundry/Anaconda Python CF Exercises'
customers = pd.read_csv(r'/Users/ECC/Career Foundry/Anaconda Python CF Exercises/02 Data/Original Data/customers.csv',index_col = False)

In [4]:
customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

## Renaming and dropping unneccesary columns.

In [25]:
customers = customers.rename(columns={'First Name': 'First_name','Surnam': 'Last_name', 'n_dependants': 'Dependents', 'fam_status': 'Marital_status'})


In [26]:
customers.head(12)

user_id First_name Last_name  Gender           STATE  Age date_joined  \
0     26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris    Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn     Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria  Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger     Casey    Male        Nebraska   45    1/1/2017   

    Dependents                    Marital_status  income  
0            3                           married  165665  
1            0                            single   59285  
2            2                           married   99568  
3            0                            single   42049  
4            1                           married   40374  
5            2                           married   49643  
6            0                            single   61746  
7            0                            single   63712  
8            1                           married  162432  
9            2                           married   32072  
10           2  living with parents and siblings   43443  
11           2                           married   47988

In [14]:
customers.shape

(206209, 10)

In [16]:
#### Contact client about if Gender, Dependents and Marital_status are required. Delete/Drop as needed.

In [15]:
customers.describe()

user_id            Age     Dependents         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [29]:
customers.dtypes

user_id            int64
First_name        object
Last_name         object
Gender            object
STATE             object
Age                int64
date_joined       object
Dependents         int64
Marital_status    object
income             int64
dtype: object

#### Data looks OK. 
## Check for mixed/incorrect types

In [31]:
for col in customers.columns.tolist():
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)
    
## Came back with First_name. Correction below. Re-running eliminated error

In [32]:
customers['First_name'] = customers['First_name'].astype('str')

In [33]:
## Check for null values
customers.isnull().sum()

user_id           0
First_name        0
Last_name         0
Gender            0
STATE             0
Age               0
date_joined       0
Dependents        0
Marital_status    0
income            0
dtype: int64

In [34]:
## No null values found.


In [36]:
## Check for duplicates 
df_dups = customers[customers.duplicated()]
df_dups


## None found

Empty DataFrame
Columns: [user_id, First_name, Last_name, Gender, STATE, Age, date_joined, Dependents, Marital_status, income]
Index: []

## Importing the order data, combine order data with customer data


In [37]:
ords_prods_merged = pd.read_pickle(r'/Users/ECC/Career Foundry/Anaconda Python CF Exercises/02 Data/Prepared Data/ords_prods_merged_21_June.pkl')

In [38]:
ords_prods_merged.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                      NaN         196                  1          0   both   
1                     15.0         196                  1          1   both   
2                     21.0         196                  1          1   both   
3                     29.0         196                  1          1   both   
4                     28.0         196                  1          1   both   
5                     19.0         196                  1          1   both   
6                     20.0         196                  1          1   both   
7                     14.0         196                  2          1   both   
8                      0.0         196                  4          1   both   
9                     30.0         196                  1          1   both   
10                     7.0         196                  2          0   both   
11                     8.0         196                  6          1   both   
12                     7.0         196                  1          1   both   
13                     7.0         196                  1          1   both   
14                    14.0         196                  1          1   both   
15                     6.0         196                  3          0   both   
16                     8.0         196                 15          1   both   
17                     6.0         196                  1          1   both   
18                     6.0         196                  2          0   both   
19                     7.0         196                 17          0   both   

    ... prices     Busiest day  Busiest_days  Busiest_period_of_day max_order  \
0   ...    9.0  Regularly busy  Regular days         Average orders        10   
1   ...    9.0  Regularly busy  Slowest days         Average orders        10   
2   ...    9.0  Regularly busy  Slowest days            Most orders        10   
3   ...    9.0      Least busy  Slowest days         Average orders        10   
4   ...    9.0      Least busy  Slowest days            Most orders        10   
5   ...    9.0  Regularly busy  Regular days         Average orders        10   
6   ...    9.0  Regularly busy  Busiest days         Average orders        10   
7   ...    9.0  Regularly busy  Busiest days            M

In [39]:
## Delete merge flag to avoid duplicate columns
del ords_prods_merged['_merge']

In [41]:
ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  prices     Busiest day  Busiest_days  \
0         Soda  ...     9.0  Regularly busy  Regular days   
1         Soda  ...     9.0  Regularly busy  Slowest days   
2         Soda  ...     9.0  Regularly busy  Slowest days   
3         Soda  ...     9.0      Least busy  Slowest days   
4         Soda  ...     9.0      Least busy  Slowest days   

  Busiest_period_of_day max_order  loyalty_flag  mean_spending spending_flag  \
0        Average orders        10  New customer           14.0  High spender   
1        Average orders        10  New customer           14.0  High spender   
2           Most orders        10  New customer           14.0  High spender   
3        Average orders        10  New customer           14.0  High spender   
4           Most orders        10  New customer           14.0  High spender   

   median_order_days            median_flag  
0               20.5  Non-frequent customer  
1               20.5  Non-frequent customer  
2               20.5  Non-frequent customer  
3               20.5  Non-frequent customer  
4               20.5  Non-frequent customer  

[5 rows x 22 columns]

In [44]:
## Checking
ords_prods_merged.shape

(32404859, 22)

In [46]:
customers.shape


(206209, 10)

## Merging ords_prods_merged with customers, using user_id as the key.
## Should end with 32404859, 32 

In [48]:
df_merge = customers.merge(ords_prods_merged, on = 'user_id', indicator = True)

In [49]:
df_merge.head(15)

user_id First_name Last_name  Gender     STATE  Age date_joined  \
0     26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
1     26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
2     26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
3     26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
4     26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
5     26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
6     26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
7     26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
8     26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
9     26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
10    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
11    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
12    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
13    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
14    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   

    Dependents Marital_status  income  ...     Busiest day  Busiest_days  \
0            3        married  165665  ...  Regularly busy  Busiest days   
1            3        married  165665  ...  Regularly busy  Regular days   
2            3        married  165665  ...  Regularly busy  Busiest days   
3            3        married  165665  ...  Regularly busy  Regular days   
4            3        married  165665  ...  Regularly busy  Slowest days   
5            3        married  165665  ...  Regularly busy  Busiest days   
6            3        married  165665  ...  Regularly busy  Busiest days   
7            3        married  165665  ...  Regularly busy  Regular days   
8            3        married  165665  ...  Regularly busy  Busiest days   
9            3        married  165665  ...  Regularly busy  Regular days   
10           3        married  165665  ...  Regularly busy  Slowest days   
11           3        married  165665  ...  Regularly busy  Busiest days   
12           3        married  165665  ...  Regularly busy  Busiest days   
13           3        married  165665  ...  Regularly busy  Busiest days   
14           3        married  165665  ...  Regularly busy  Regular days   

    Busiest_period_of_day  max_order  loyalty_flag  mean_spending  \
0          Average orders          8  New customer           12.6   
1             Most orders          8  New customer           12.6   
2             Most orders          8  New customer           12.6   
3          Average orders          8  New customer           12.6   
4             Most orders          8  New customer           12.6   
5          Average orders          8  New customer           12.6   
6          Average orders          8  New customer           12.6   
7             Most orders          8  New customer           12.6   
8             Most orders          8  New customer           12.6   
9          Average orders          8  New customer           12.6   
10            Most orders          8  New customer           12.6   
11         Average orders          8  New customer           12.6   
12         Average orders          8  New customer           12.6   
13         Average orders          8  New customer           12.6   
14            Most orders          8  New customer           12.6   

    spending_flag  median_order_days median_flag  _merge  
0    High spender               19.0         NaN    both  
1    High spender               19.0         NaN    both  
2    High spender               19.0         NaN    both  
3    High spender               19.0         NaN    both  
4    High spender               19.0         NaN    both  
5    High spender               19.0         NaN    both  
6    High spender               19.0         NaN    both  
7    High spender               19.0         NaN    both  
8    High spen

In [50]:
df_merge['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

In [51]:
df_merge.shape

(32404859, 32)

In [57]:
## Great success. Saving, exporting and moving on to Task 2.

df_merge.to_pickle(os.path.join(r'/Users/ECC/Career Foundry/Anaconda Python CF Exercises/02 Data/Prepared Data/ords_cust_merged.pkl'))